Source/RimConnection/Managers/PawnCreationManager.cs

```diff
@@ -69,10 +95,32 @@ public static List<Thing> generateAwfulColonists(int amount, string name = null)
                 var newPawn = PawnGenerator.GeneratePawn(PawnKindDefOf.Colonist);
                 newPawn.SetFaction(Faction.OfPlayer);
 
-                // fuck up their traits
-                var randomTraits = badTraits.InRandomOrder().Take(3);
+                // Create a dictionary to store available traits for each trait type
+                Dictionary<TraitDef, List<Trait>> availableTraits = new Dictionary<TraitDef, List<Trait>>();
+                foreach (Trait trait in badTraits)
+                {
+                    if (!availableTraits.ContainsKey(trait.def))
+                    {
+                        availableTraits[trait.def] = new List<Trait>();
+                    }
+                    availableTraits[trait.def].Add(trait);
+                }
+
+                List<Trait> selectedTraits = new List<Trait>();
+
+                // Randomly select 3 different trait types
+                while (selectedTraits.Count < 3 && availableTraits.Count > 0)
+                {
+                    TraitDef randomTraitDef = availableTraits.Keys.RandomElement();
+                    List<Trait> traitsForType = availableTraits[randomTraitDef];
+                    Trait selectedTrait = traitsForType.RandomElement();
+                    selectedTraits.Add(selectedTrait);
+                    availableTraits.Remove(randomTraitDef);
+                }
```

Comment:
This code is basically repeated for each generation with the only things changing being what list it's pulling from and how many it's pulling. Can we make a function that takes those 2 arguments and returns us a list of traitDefs too apply to the colonist?